db/db_impl/db_impl.cc

```diff
@@ -1061,6 +1061,204 @@ FSDirectory* DBImpl::GetDataDir(ColumnFamilyData* cfd, size_t path_id) const {
+                                         ApplyReplicationLogRecordInfo* info) {
+  JobContext job_context(0, true);
+  Status s;
+
+  {
+    WriteThread::Writer w;
+    InstrumentedMutexLock l(&mutex_);
+    write_thread_.EnterUnbatched(&w, &mutex_);
+
+    switch (record.type) {
+      case ReplicationLogRecord::kMemtableWrite: {
+        mutex_.Unlock();
+        WriteBatch batch;
+        s = WriteBatchInternal::SetContents(&batch, std::move(record.contents));
+        assert(s.ok());
+
+        if (versions_->LastSequence() + 1 !=
+            WriteBatchInternal::Sequence(&batch)) {
+          std::ostringstream oss;
+          oss << "Gap in sequence numbers, expected="
+              << versions_->LastSequence() + 1
+              << " got=" << WriteBatchInternal::Sequence(&batch);
+          return Status::Corruption(oss.str());
+        }
+
+        SequenceNumber next_seq;
+        s = WriteBatchInternal::InsertInto(
+            &batch, column_family_memtables_.get(), &flush_scheduler_,
+            &trim_history_scheduler_,
+            true /* ignore_missing_column_families_ */, 0 /* log_number */,
+            this, false /* concurrent_memtable_writes */, &next_seq,
+            nullptr /* has_valid_writes */, seq_per_batch_, batch_per_txn_);
+        if (s.ok()) {
+          versions_->SetLastSequence(next_seq - 1);
+        }
+        mutex_.Lock();
+        break;
+      }
+      case ReplicationLogRecord::kMemtableSwitch: {
+        WriteContext write_context;
+        autovector<ColumnFamilyData*> cfds;
+        SelectColumnFamiliesForAtomicFlush(&cfds);
+        for (const auto cfd : cfds) {
+          cfd->Ref();
+          s = SwitchMemtable(cfd, &write_context, "");
+          cfd->UnrefAndTryDelete();
+          if (!s.ok()) {
+            break;
+          }
+        }
+        break;
+      }
+      case ReplicationLogRecord::kManifestWrite: {
+        Slice src = record.contents;
+        autovector<VersionEdit> edits;
+        DeserializeReplicationLogManifestWrite(&src, &edits);
+
+        auto mutable_options =
+            default_cf_handle_->cfd()->GetLatestMutableCFOptions();
+        std::unique_ptr<ColumnFamilyOptions> cf_options;
+
+        autovector<ColumnFamilyData*> cfds;
+        autovector<const MutableCFOptions*> mutable_cf_options_list;
+        autovector<autovector<VersionEdit*>> edit_lists;
+
+        std::vector<uint32_t> added_column_families;
+
+        auto current_update_sequence = versions_->GetManifestUpdateSequence();
+        uint64_t latest_applied_update_sequence = 0;
+        for (auto& e : edits) {
+          if (!e.HasManifestUpdateSequence()) {
+            return Status::InvalidArgument(
+                "Manifest write doesn't have a ManifestUpdateSequence");
+          }
+          latest_applied_update_sequence = e.GetManifestUpdateSequence();
+          if (e.GetManifestUpdateSequence() <= current_update_sequence) {
```

Comment:
Currently, seems for `ManifestUpdatesequence,` we don't include epoch but just the sequence number. I'm thinking about following corner case, in which including the epoch could be helpful:

Suppose we have A,B,C,D with A being the leader and the rest being followers.

After network partition, C,D can't connect to A, so we have two separated Leader Follower groups:

[A, B], both of them think A is the leader
[C,D], after leader election, C becomes the real leader.

But what if network partition is a transient issue and everything recovers after leader election:

Case 1) We keep the same state in normal case
Case 2) But if A, B tries to rejoin the group [C,D], since the log between the two groups already diverges, it doesn't make sense for A,B to continue consuming from C and D. But if we only compare the sequence number here, there is a small chance that the manifest update sequence in group [A,B] >= then largest manifest update sequence in [C,D] and therefore rejoining the [C,D] group.

Reply:
Discussed offline, the replication layer will have the ability to detect divergence and reject A and B from rejoining the C, D group.